File: scripts/goodshort-scraper/src/upload-to-r2.ts
/**
 * Upload Downloaded Videos to Cloudflare R2
 * 
 * Scans downloads/ folder and uploads all MP4s and covers to R2
 * Generates metadata JSON with public R2 URLs
 * 
 * Usage: npx ts-node src/upload-to-r2.ts
 */

import { S3Client, PutObjectCommand, HeadObjectCommand } from '@aws-sdk/client-s3';
import * as fs from 'fs';
import * as path from 'path';
import * as dotenv from 'dotenv';

dotenv.config();

// R2 Configuration
const R2_CONFIG = {
    accountId: process.env.R2_ACCOUNT_ID!,
    accessKeyId: process.env.R2_ACCESS_KEY_ID!,
    secretAccessKey: process.env.R2_SECRET_ACCESS_KEY!,
    bucketName: process.env.R2_BUCKET_NAME || 'kingshortid',
    publicUrl: process.env.R2_PUBLIC_URL!,
};

// Initialize S3 client for R2
const r2Client = new S3Client({
    region: 'auto',
    endpoint: `https://${R2_CONFIG.accountId}.r2.cloudflarestorage.com`,
    credentials: {
        accessKeyId: R2_CONFIG.accessKeyId,
        secretAccessKey: R2_CONFIG.secretAccessKey,
    },
});

interface UploadResult {
    bookId: string;
    coverUrl: string | null;
    episodes: {
        chapterId: string;
        videoUrl: string;
        fileSize: number;
    }[];
}

/**
 * Check if file already exists in R2
 */
async function fileExistsInR2(key: string): Promise<boolean> {
    try {
        await r2Client.send(new HeadObjectCommand({
            Bucket: R2_CONFIG.bucketName,
            Key: key,
        }));
        return true;
    } catch (error: any) {
        if (error.name === 'NotFound') {
            return false;
        }
        throw error;
    }
}

/**
 * Upload file to R2
 */
async function uploadToR2(
    filePath: string,
    key: string,
    contentType: string
): Promise<string> {
    const fileBuffer = fs.readFileSync(filePath);

    // Check if already uploaded (skip to save time)
    const exists = await fileExistsInR2(key);
    if (exists) {
        console.log(`    ⏭️  Already uploaded: ${key}`);
        return `${R2_CONFIG.publicUrl}/${key}`;
    }

    await r2Client.send(new PutObjectCommand({
        Bucket: R2_CONFIG.bucketName,
        Key: key,
        Body: fileBuffer,
        ContentType: contentType,
    }));

    const publicUrl = `${R2_CONFIG.publicUrl}/${key}`;
    console.log(`    ✅ Uploaded: ${key}`);
    return publicUrl;
}

/**
 * Process a single drama folder
 */
async function processDrama(dramaPath: string): Promise<UploadResult> {
    const bookId = path.basename(dramaPath);

    console.log(`\n${'='.repeat(50)}`);
    console.log(`Processing Drama: ${bookId}`);
    console.log(`${'='.repeat(50)}`);

    const result: UploadResult = {
        bookId,
        coverUrl: null,
        episodes: [],
    };

    // Upload cover if exists
    const coverPath = path.join(dramaPath, 'cover.jpg');
    if (fs.existsSync(coverPath)) {
        const coverKey = `goodshort-content/${bookId}/cover.jpg`;
        result.coverUrl = await uploadToR2(coverPath, coverKey, 'image/jpeg');
    }

    // Upload all episode MP4s
    const episodeFiles = fs.readdirSync(dramaPath)
        .filter(file => file.startsWith('episode_') && file.endsWith('.mp4'))
        .sort();

    console.log(`  Found ${episodeFiles.length} episodes to upload`);

    for (const episodeFile of episodeFiles) {
        const episodePath = path.join(dramaPath, episodeFile);
        const chapterId = episodeFile.match(/episode_(\d+)\.mp4/)?.[1];

        if (!chapterId) {
            console.log(`  ⚠️  Skipping invalid filename: ${episodeFile}`);
            continue;
        }

        const videoKey = `goodshort-content/${bookId}/${chapterId}.mp4`;
        const videoUrl = await uploadToR2(episodePath, videoKey, 'video/mp4');

        const fileSize = fs.statSync(episodePath).size;

        result.episodes.push({
            chapterId,
            videoUrl,
            fileSize,
        });
    }

    console.log(`  ✓ Uploaded ${result.episodes.length} episodes + cover`);

    return result;
}

/**
 * Main upload function
 */
async function main() {
    console.log('\n' + '='.repeat(60));
    console.log('GoodShort R2 Upload Tool');
    console.log('='.repeat(60));

    // Validate environment
    if (!R2_CONFIG.accountId || !R2_CONFIG.accessKeyId || !R2_CONFIG.secretAccessKey || !R2_CONFIG.publicUrl) {
        console.error('\n❌ Missing R2 configuration in .env:');
        console.error('   R2_ACCOUNT_ID');
        console.error('   R2_ACCESS_KEY_ID');
        console.error('   R2_SECRET_ACCESS_KEY');
        console.error('   R2_PUBLIC_URL');
        process.exit(1);
    }

    const downloadsDir = path.join(__dirname, '..', 'downloads');

    if (!fs.existsSync(downloadsDir)) {
        console.error('\n❌ downloads/ folder not found');
        console.error('   Run batch-download.ts first to download videos');
        process.exit(1);
    }

    // Find all drama folders
    const dramaFolders = fs.readdirSync(downloadsDir)
        .map(name => path.join(downloadsDir, name))
        .filter(p => fs.statSync(p).isDirectory());

    console.log(`\nFound ${dramaFolders.length} drama folders to process\n`);

    const results: UploadResult[] = [];

    for (const dramaPath of dramaFolders) {
        try {
            const result = await processDrama(dramaPath);
            results.push(result);
        } catch (error: any) {
            console.error(`\n❌ Failed to upload drama ${path.basename(dramaPath)}:`, error.message);
        }
    }

    // Save upload results to JSON
    const outputFile = path.join(__dirname, '..', 'r2-upload-results.json');
    fs.writeFileSync(outputFile, JSON.stringify({
        uploadedAt: new Date().toISOString(),
        totalDramas: results.length,
        totalEpisodes: results.reduce((sum, r) => sum + r.episodes.length, 0),
        dramas: results,
    }, null, 2));

    console.log('\n' + '='.repeat(60));
    console.log('✅ Upload Complete!');
    console.log('='.repeat(60));
    console.log(`Total Dramas: ${results.length}`);
    console.log(`Total Episodes: ${results.reduce((sum, r) => sum + r.episodes.length, 0)}`);
    console.log(`\n📄 Results saved to: ${outputFile}`);
    console.log('\nNext step: Run import-to-kingshortid.ts to add to database');
    console.log('='.repeat(60) + '\n');
}

main().catch(console.error);

File: scripts/goodshort-scraper/upload_to_r2.py
#!/usr/bin/env python3
"""
R2 UPLOADER - Upload GoodShort Data to Cloudflare R2
===================================================

Uploads scraped drama data to R2 bucket 'kingshortid'

Required env vars:
- R2_ACCOUNT_ID
- R2_ACCESS_KEY_ID
- R2_SECRET_ACCESS_KEY

Usage:
    python upload_to_r2.py
"""

import boto3
import json
import os
from pathlib import Path
from typing import List, Dict
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configuration
SCRIPT_DIR = Path(__file__).parent
SOURCE_DIR = SCRIPT_DIR / "r2_ready"  # Updated to match our folder structure
R2_BUCKET = os.getenv("R2_BUCKET_NAME", "kingshort")

# R2 Credentials - Extract from .env
R2_ENDPOINT = os.getenv("R2_ENDPOINT", "")
R2_ACCESS_KEY_ID = os.getenv("R2_ACCESS_KEY_ID", "")
R2_SECRET_ACCESS_KEY = os.getenv("R2_SECRET_ACCESS_KEY", "")

# Extract account ID from endpoint URL
# Format: https://{ACCOUNT_ID}.r2.cloudflarestorage.com
if R2_ENDPOINT:
    R2_ACCOUNT_ID = R2_ENDPOINT.split('://')[1].split('.')[0]
else:
    R2_ACCOUNT_ID = ""

class R2Uploader:
    """Upload drama data to Cloudflare R2"""
    
    def __init__(self):
        self.uploaded_files = 0
        self.total_bytes = 0
        self.dramas_uploaded = []
        
        # Initialize S3 client for R2
        self.s3_client = boto3.client(
            's3',
            endpoint_url=f'https://{R2_ACCOUNT_ID}.r2.cloudflarestorage.com',
            aws_access_key_id=R2_ACCESS_KEY_ID,
            aws_secret_access_key=R2_SECRET_ACCESS_KEY,
            region_name='auto'
        )
    
    def get_content_type(self, file_path: Path) -> str:
        """Get content type for file"""
        ext = file_path.suffix.lower()
        content_types = {
            '.jpg': 'image/jpeg',
            '.jpeg': 'image/jpeg',
            '.png': 'image/png',
            '.json': 'application/json',
            '.m3u8': 'application/vnd.apple.mpegurl',
            '.ts': 'video/mp2t'
        }
        return content_types.get(ext, 'application/octet-stream')
    
    def upload_file(self, local_path: Path, s3_key: str) -> bool:
        """Upload single file to R2"""
        try:
            content_type = self.get_content_type(local_path)
            
            # Upload with public-read ACL
            self.s3_client.upload_file(
                str(local_path),
                R2_BUCKET,
                s3_key,
                ExtraArgs={
                    'ContentType': content_type,
                    'ACL': 'public-read'  # Make files publicly accessible
                }
            )
            
            file_size = local_path.stat().st_size
            self.uploaded_files += 1
            self.total_bytes += file_size
            
            return True
            
        except Exception as e:
            print(f"    ❌ Upload failed: {e}")
            return False
    
    def upload_drama(self, drama_folder: Path) -> Dict:
        """Upload complete drama folder to R2 (including episode subfolders)"""
        drama_id = drama_folder.name
        upload_result = {
            "drama_id": drama_id,
            "files_uploaded": 0,
            "errors": []
        }
        
        # Recursively upload all files and folders
        for item in drama_folder.rglob('*'):
            if not item.is_file():
                continue
            
            # Generate R2 path (preserve folder structure)
            relative_path = item.relative_to(drama_folder)
            s3_key = f"{drama_id}/{relative_path.as_posix()}"
            
            print(f"    📤 {relative_path}...")
            
            if self.upload_file(item, s3_key):
                upload_result["files_uploaded"] += 1
            else:
                upload_result["errors"].append(str(relative_path))
        
        return upload_result
    
    def upload_all(self):
        """Upload all dramas from r2_upload_ready"""
        if not SOURCE_DIR.exists():
            print(f"❌ Source directory not found: {SOURCE_DIR}")
            print(f"Run aggressive_scrape_to_r2.py first!")
            return
        
        # Find all drama folders
        drama_folders = [f for f in SOURCE_DIR.iterdir() if f.is_dir()]
        
        if not drama_folders:
            print(f"❌ No drama folders found in {SOURCE_DIR}")
            return
        
        print(f"✅ Found {len(drama_folders)} dramas to upload\n")
        
        # Upload each drama
        for i, folder in enumerate(drama_folders, 1):
            print(f"{'='*70}")
            print(f"📤 Uploading {i}/{len(drama_folders)}: {folder.name}")
            print(f"{'='*70}")
            
            result = self.upload_drama(folder)
            
            if result["files_uploaded"] > 0:
                print(f"  ✅ Uploaded {result['files_uploaded']} files")
                self.dramas_uploaded.append(result["drama_id"])
            
            if result["errors"]:
                print(f"  ⚠️  Errors: {len(result['errors'])} files failed")
            
            print()
        
        # Summary
        print(f"\n{'='*70}")
        print(f"✅ R2 UPLOAD COMPLETE!")
        print(f"{'='*70}")
        print(f"\n📊 Statistics:")
        print(f"  - Dramas uploaded: {len(self.dramas_uploaded)}")
        print(f"  - Files uploaded: {self.uploaded_files}")
        print(f"  - Total size: {self.total_bytes / 1024 / 1024:.2f} MB")
        print(f"\n📦 R2 Bucket: {R2_BUCKET}")
        print(f"🌐 Path: goodshort/{{drama_id}}/")
        print(f"\n💡 Public URLs:")
        for drama_id in self.dramas_uploaded:
            print(f"  https://pub-YOUR_ID.r2.dev/goodshort/{drama_id}/cover.jpg")
            print(f"  https://pub-YOUR_ID.r2.dev/goodshort/{drama_id}/metadata.json")
        print()

def check_credentials():
    """Check if R2 credentials are configured"""
    if not R2_ACCOUNT_ID or not R2_ACCESS_KEY_ID or not R2_SECRET_ACCESS_KEY:
        print("❌ R2 credentials not configured!")
        print("\nPlease set environment variables:")
        print("  - R2_ACCOUNT_ID")
        print("  - R2_ACCESS_KEY_ID")
        print("  - R2_SECRET_ACCESS_KEY")
        print("\nOr create .env file in script directory:")
        print("")
        print("R2_ACCOUNT_ID=your_account_id")
        print("R2_ACCESS_KEY_ID=your_access_key")
        print("R2_SECRET_ACCESS_KEY=your_secret_key")
        print("")
        return False
    return True

def main():
    print("\n" + "="*70)
    print("☁️  R2 UPLOADER - GoodShort to Cloudflare R2")
    print("="*70 + "\n")
    
    # Check credentials
    if not check_credentials():
        return
    
    uploader = R2Uploader()
    uploader.upload_all()

if __name__ == "__main__":
    main()
